pkg/app/server.go

```diff
@@ -145,6 +149,24 @@ func RunKubeStateMetrics(ctx context.Context, opts *options.Options) error {
 		}
 	}
 
+	kubeConfig, err := clientcmd.BuildConfigFromFlags(opts.Apiserver, opts.Kubeconfig)
+	if err != nil {
+		return fmt.Errorf("failed to build config from flags: %v", err)
+	}
+
+	// Ensure cache prerequisites are set up only once.
+	if gvkMaps.Map == nil {
+		discoveryClient := cgodiscovery.NewDiscoveryClientForConfigOrDie(kubeConfig)
+
+		// Start an independent goroutine to periodically update `gvkMaps`.
+		t := time.NewTicker(discovery.Interval)
+		go func() {
+			for range t.C {
+				gvkMaps.Discover(discoveryClient)
+			}
+		}()
```

Comment:
this need to be cancellable via the context

Reply:
IIUC do we want to cancel discovering when KSM is reloaded?

Reply:
why we would want both to be separated? I am thinking that discovery should be an integral part of the ksm server. It ksm is being cleaned-up/cancelled, I would expect the same for all the goroutines it starts

Reply:
ACK, I initially thought about keeping it separate so that the discovery caches are unaffected by the parent process' state, but this makes sense. I've updated the code accordingly.